design/open/template-server/template-server.md

```diff
@@ -0,0 +1,130 @@
+
+If storing on lakeFS, IAM authorizes users to see this template by
+`fs:ReadObject`, at which point expansion can start.  If we want to store
+templates on the blockstore types, we will probably need to add a new IAM
+action type for template expansion, or alternatively overload
+`fs:ReadObject` to support off-lakeFS URIs.  We will supply the following
+replacements (at least):
+
+| template function | value source                        | IAM permissions                              |
+|:------------------|:------------------------------------|:---------------------------------------------|
+| config            | exportable lakeFS configuration[^1] | `arn:lakefs:template:::config/path/to/field` |
+| object            | (small) object on lakeFS            | IAM read permission for that object          |
+| contenttype       | none; set `Content-Type:`           | (none)                                       |
+| new_credentials   | new credentials added to user       | `auth:CreateCredentials`                     |
+
+A configuration variable will be "exportable" if its struct field is tagged
+`` `export:yes` ``, or (if implementing that is too hard) if its type
+implements
+
+```go
+interface Exportable {
+	Export() string
+}
+```
+
+IAM is still checked for exportable configuration variables (which will be
+ON by default for users, but can be removed).
+
+Adding a freeform dictionary to our config will allow admins to set up
+any needed configuration.
+
+We shall supply the user object in `.user` and the parsed query args in
+`.query`[^2], to allow conditional operations.  We will probably add new
+functions or objects.
+
+## Template expansion flow
+
+#### _:warning: This flow assumes templates stored on lakeFS. :warning:_
+
+1. User accesses the template
+   http://<lakefs>/api/v1/templates/main/expand/me.
+1. lakeFS expands to a lakeFS path (by default this is
+   lakefs://lakefs-templates/main/expand/me).
+1. lakeFS uses IAM to verify user has `fs:ReadObject` permission on this
+   object.
+1. Object is parsed into a text/template.
+1. lakeFS creates template
+   [`Funcs`](https://pkg.go.dev/text/template#FuncMap) that will check IAM
+   permissions as required for expansion.
+1. Template gets expanded and returned to user.
+
+### Credentials
+
+We supply a function `new_credentials` that creates a new set of credential
+for the user, registers them, and returns them.  See this [example of template
+functions returning multiple values][example_template_multiple].
+
+This lets us implement the requirement to create and return new credentials
+as part of a downloaded configuration.
+
+## Examples
+
+(All examples subject to change as we define object format, and decide
+what functions are allowed.)
+
+Set a template that looks like this:
+
+```conf
+spark.hadoop.fs.lakefs.impl=io.lakefs.LakeFSFileSystem
+{{with $creds := new_credentials}}
+spark.hadoop.fs.lakefs.access_key={{$creds.ID}}
```

Comment:
```suggestion
spark.hadoop.fs.lakefs.access_key={{$creds.KEY}}
```
?

Reply:
Not done: it's called the "ID" everywhere, in AWS and in our docs.